File: problems/110. Balanced Binary Tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        int height = getHeight(root);
        return (height != -1);
    }

    private int getHeight(TreeNode root) {
        if (root == null) return 0;
        int leftHeight = root.left == null ? 0 : getHeight(root.left);
        int rightHeight = root.right == null ? 0 : getHeight(root.right);
        if (leftHeight == -1 || rightHeight == -1) return -1;
        else if (Math.abs(leftHeight - rightHeight) > 1) return -1;
        else return Math.max(leftHeight, rightHeight) + 1;
    }
}

File: problems/110. Balanced Binary Tree/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        int height = getHeight(root);
        return (height != -1);
    }

    int getHeight(TreeNode* root) {
        if (!root) return 0;
        int leftHeight = root->left ? getHeight(root->left) : 0;
        int rightHeight = root->right ? getHeight(root->right) : 0;
        if (leftHeight == -1 || rightHeight == -1) return -1;
        else if (std::abs(getHeight(root->left) - getHeight(root->right)) > 1) return -1;
        else return max(leftHeight, rightHeight) + 1;
    }
};